lib/Mojo/Date.pm

```diff
@@ -54,16 +54,29 @@ sub parse {
 sub to_datetime {
 
   # RFC 3339 (1994-11-06T08:49:37Z)
-  my ($s, $m, $h, $day, $month, $year) = gmtime(my $epoch = shift->epoch);
+  my ($s, $m, $h, $day, $month, $year) = _gmftime(shift->epoch);
   my $str = sprintf '%04d-%02d-%02dT%02d:%02d:%02d', $year + 1900, $month + 1,
     $day, $h, $m, $s;
-  return $str . ($epoch =~ /(\.\d+)$/ ? $1 : '') . 'Z';
+  return $str . ($s =~ /(\.\d+)$/ ? $1 : '') . 'Z';
 }
 
+sub _gmftime {
+  my $epoch = shift;
+  if ($epoch =~ /^(-?\d+)(\.\d+)$/) {
+    my $t = ($epoch < 0) ? $1 - 1      : $1;    # floor
+    my $f = ($epoch < 0) ? $epoch - $t : $2;
+    my @t = gmtime($t);
+    $t[0] += $f;
+    return @t;
+  }
+  return gmtime($epoch);
+}
```

Comment:
Well played. =) I like the way you compressed the most common case to an efficient expression at the top of the function and the combination of `@t` and `$f` in a single expression.

I am inclined to be a little more verbose to highlight that the important thing here is to decompose `$epoch` into `$t + $f` where `0 ≤ $f < 1`. Something like

```perl
sub _gmftime {
  return &gmtime unless $_[0] =~ /^(-?\d+)(\.\d+)$/;
  # decompose $epoch into $t + $f, where 0 < $f < 1
  my $t = ($_[0] < 0) ? $1 - 1      : $1;
  my $f = ($_[0] < 0) ? $_[0] - $t : $2;
  @t = gmtime $t;
  return shift(@t) + $f, @t;
}
```

Notice this has one less floating point operation (when you compute `$1 -1` twice at `$_[0] - $1 - 1` and `$1 -1`).

Also the variant below can avoid an extra floating point operation above that is caused by the numeric comparison `$_[0] < 0`.

```perl
sub _gmftime {
  return &gmtime unless $_[0] =~ /^((-?)\d+)(\.\d+)$/;
  # decompose $epoch into $t + $f, where 0 < $f < 1
  my $t = $2 ? $1 - 1      : $1;
  my $f = $2 ? $_[0] - $t : $3;
  @t = gmtime $t;
  return shift(@t) + $f, @t;
}
```
Does it look worth?